Dev/define cmd

thattil opened the pull request:
New:
- commands: define_reg, define_bit, definelist_reg, definelist_bit

Definitions can only be 32 characters long and is stored in shared memory only for ctb and xilinx ctb.

Define_reg: Defines a register address
   - examples:
              sls_detector_put define_reg test_reg 0x200
              sls_detector_get define_reg test_reg # => 0x200
              sls_detector_get define_reg 0x200 # => test_reg

              sls_detector_put reg test_reg 0x1
              sls_detector_put setbit test_reg 2
              sls_detector_put clearbit test_reg 2
              sls_detector_get getbit test_reg 2


Define_bit: Defines a bit position and register address together
- examples:
              sls_detector_put define_bit test_bit test_reg 2
              sls_detector_put define_bit test_bit 0x200 2

              sls_detector_get define_bit test_bit # => test_reg 2
              sls_detector_get define_bit test_reg 2 # =>test_bit
              sls_detector_get define_bit 0x200 2 # => test_bit

              sls_detector_put setbit test_bit
              sls_detector_put clearbit test_bit
              sls_detector_get getbit test_bit

All of them also support values as before. 

- Had to move these from commands.yaml to cmdspecial to be able to allow parsing by string (defined names) or integer (original method).

address and bit names are only for ctb and xilinx=> written to CtbConfig shared memory. Change shm date.
API added to set them and get them, and get the list.

thattil committed:
wip

thattil committed:
basic ctb config api for register and bit names

thattil committed:
wip

thattil committed:
wip

thattil committed:
wip

thattil committed:
wip

thattil committed:
Merge branch 'developer' into dev/define_cmd

thattil committed:
tests for define and definelist pass. yet to implement using them for reg, setbit, clearbit and getbit

thattil committed:
refactor test, wip

thattil committed:
wip, reg define

thattil committed:
wip, works for read and write reg with pipes

thattil committed:
done with tests

thattil committed:
works

thattil committed:
formatting

thattil committed:
improved autocomplete for getbit,setbit, clearbit

thattil committed:
validate autocomplete

thattil committed:
definelist has no put

thattil committed:
updating help

thattil committed:
refactor

thattil committed:
refactor

thattil committed:
refactor define cmdline

thattil committed:
refactor

erikfrojdh commented on slsDetectorSoftware/include/sls/Detector.h:
```diff
@@ -1837,6 +1837,53 @@ class Detector {
+
+    /** [CTB][Xilinx CTB] */
+    void
+    setBitDefinitions(const std::vector<std::pair<std::string, int>> &list);
```
same as the above a std::map<string, int>?

thattil replied:
Will do

erikfrojdh commented on slsDetectorSoftware/include/sls/Detector.h:
```diff
@@ -1837,6 +1837,53 @@ class Detector {
+        const std::vector<std::pair<std::string, int>> &list);
+
+    /** [CTB][Xilinx CTB] */
+    std::vector<std::pair<std::string, int>> getRegisterDefinitions() const;
```
Vector or std::map<string, int>

thattil replied:
i guess you mean map? cant do vector<string, int>. Ill implement that anyway.

erikfrojdh commented on slsDetectorSoftware/src/CtbConfig.cpp:
```diff
@@ -238,4 +238,176 @@ std::vector<std::string> CtbConfig::getSlowADCNames() const {
 
 const char *CtbConfig::shm_tag() { return shm_tag_; }
 
+std::optional<Entry *> CtbConfig::findEntryByName(const std::string &name,
```
My gut feeling is that the lookup would be much easier with map/unordered map

erikfrojdh reviewed (commented):
Think about the implementation.

thattil committed:
converting char array+int in runtimeerror compiles but throws at runtime.Fixed.Tested for it. Also check if string or int before using getregisterdefinitonbyvalue to see if it threw to call the other function. because both of it can throw and we should differentiate the issues for both

thattil committed:
removed std::vector<std::pair<string,int> to std::map<string, int> for defiitions list

thattil committed:
Dev/define cmd tie bit to reg (#1328)


* strong type wip

* moved everythign to bit_utils class

* wip, still on testing in ctb reg

* passed tests for define

thattil committed:
wip for pybindings

thattil committed:
python compiles

thattil committed:
added tests for python

erikfrojdh committed:
Merge branch 'developer' into dev/define_cmd

thattil committed:
Merge branch 'developer' into dev/define_cmd

AliceMazzoleni99 commented on python/slsdet/utils.py:
```diff
@@ -141,21 +141,38 @@ def make_ip(arg):
-        return tuple(transform(a) for a in arg)
+        # special case for BitPosition
+        if transform is _slsdet.BitPosition:
+            addr, bit = arg
```
why only for transform _slsdet.BitPosition?

thattil replied:
Because Bit Position has 2 values unlike the others..

AliceMazzoleni99 replied:
ah got it.

AliceMazzoleni99 commented on python/src/detector.cpp:
```diff
@@ -1969,29 +2045,55 @@ void init_det(py::module &m) {
+        (void (Detector::*)(sls::BitPosition, bool, sls::Positions)) &
+            Detector::setBit,
+        py::arg(), py::arg() = false, py::arg() = Positions{});
     CppDetectorApi.def(
```
Do we want to enable both strong types and the non strong types?

thattil replied:
In Detector.h, we are keeping both to make it backward compatible. Did we decide this in a prior meeting? If not, maybe we should discuss this. Python bindings if it should stay the same as Detector class. We should also discuss if it hasnt been.

thattil replied:
We keep deprecated functions (not strong typed arguments) in Detector cpp class until a major version where we would break backwards compatibility. About the warning in python bindings because of it, we still need to decide how to handle it.

AliceMazzoleni99 commented on python/src/detector_in.cpp:
```diff
@@ -5,8 +5,10 @@
 #include "sls/sls_detector_defs.h"
 #include "sls/TimeHelper.h"
+#include "sls/bit_utils.h"
 
```
only keep once

thattil replied:
Done!

thattil committed:
removed duplicates

AliceMazzoleni99 commented on slsDetectorSoftware/generator/commands.yaml:
```diff
@@ -2761,6 +2761,66 @@ sleep:
             - argc: 2
               arg_types: [ int, special::time_unit ]
 
```
what is this used for?

AliceMazzoleni99 replied:
is there no help message?

thattil replied:
I think this is used for the infer script to infer if the command is for sls_detector_put or sls_detector_get. This command (and all others with "is_description: true") is implemented in CallerSpecial.cpp. And writing it here only helps to infer the put or get. This is what I understand.

AliceMazzoleni99 commented on slsDetectorSoftware/generator/extended_commands.yaml:
I didnt look over this part

thattil replied:
The extended commands.yaml you dont really need to, as it is auto generated. Upto you.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        WrongNumberOfParameters(2);
+    }
+    std::string mode = args[0];
+    if (mode == "addr") {
```
is it addr or define addr not clear from helper string

thattil replied:
'define' is the command.. 'addr' or 'bit' is the mode.. There are examples in the helper string.

\n\tsls_detector_put define addr test_reg 0x200"
              "\n\tsls_detector_put define bit test_bit test_reg 2"
              "\n\tsls_detector_put define bit test_bit 0x200 2"

And many more. Maybe we can discuss tomorrow how I can improve this to be more readable. 

I also see that I didnt update the PR description to reflect the change to "addr" naming. Ill fix this now.

AliceMazzoleni99 replied:
Maybe cretae two functions define_reg and define_bit

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        auto det_type = det->getDetectorType().squash();
+        if (det_type != defs::XILINX_CHIPTESTBOARD &&
+            det_type != defs::CHIPTESTBOARD) {
+            throw RuntimeError("Could not parse bit name " + addr_or_bitname +
```
why is this error message only here?

thattil replied:
I didnt understand this question. This error message is if you try to use bit name or address name definitions as they are only for ctb and xilinx ctb. They are in other places as well.

AliceMazzoleni99 replied:
Oke got it. Sorry misunderstood.

AliceMazzoleni99 commented on python/slsdet/utils.py:
```diff
@@ -141,21 +141,38 @@ def make_ip(arg):
+
+def make_register_value(arg):
+    return _make(arg, _slsdet.RegisterValue)
 
```
use it in detector.py

AliceMazzoleni99 replied:
Or actually its never used right now I dont think we use it anyway.

thattil replied:
I think we discussed this yesterday. That when a user uses it, they should use it to convert it to RegisterValue to use the non deprecated functions.  Right now, both are possible.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+              "\n\teg."
+              "\n\tsls_detector_put define addr test_reg 0x200"
+              "\n\tsls_detector_put define bit test_bit test_reg 2"
+              "\n\tsls_detector_put define bit test_bit 0x200 2"
```
as this can still be mixed up add an info Defining test_bit as bit 2 of register with address 0x200

AliceMazzoleni99 replied:
Maybe put it in Detector::setBitPosition then the info is also shown when called from python

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+            // get address from name
+            else {
+                auto t = det->getRegisterDefinition(args[1]);
+                os << t.str() << '\n';
```
mmh i think one should return the same thing twice. e.g. Detector::getRegisterDefinition always returns a std::map<std::string, RegisterAdress>

thattil replied:
Why should it return a map when there is only a single value required here... I believe if one needs to just get a single registeraddress from a name, then it needn't be more complex and just return the string. Or have I misunderstood you?

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+              "is tied to a specific register already.\n"
+              "\n\n\teg."
+              "\n\tsls_detector_get reg test_reg"
+              "\n\tsls_detector_put reg test_bit"
```
should this be sls_detector_get

thattil replied:
No, setting just the bit will set only that bit in that register.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        if (!is_hex_or_dec_uint(sval)) {
+            throw RuntimeError(
+                "If " + sval +
+                " is a bit name, use only one argument to specify it.");
```
why bit name? Dont get the error message

thattil replied:
If you give 2 arguments.. 0x200 test_bit... test_bit is already connected to a register.. so you need to give only test_bit..

AliceMazzoleni99 replied:
But why is this in parseRegisterAdress and not parseBitPoistion? Maybe I misunderstand what parseRgeisterAdress should do. Or when you call reg["bit_name"] does it give you the value at the bit poistion? or do you have to use getbit?

thattil replied:
It is similar but the name is not "parseRegisterAddress". This function only passes a register address. 

However, the method we are discussing is 'parseRegisterAddressAndValue'. It is called when you "write" a value to register. This can currently be either a 
- combination of an address (value or name) and a hardcoded value, or
- concatenation of bit names (here you do not require an address as bit names are already connected to an address)

thattil replied:
Removing bit namees and concatenation in reg command to avoid misinterpretations. So this function will be removed.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+    }
+    // parse bit name (or concatenation of them)
+    else if (argsSize == 1) {
+        std::string sbitNames = args[0];
```
is it the register or bit?

thattil replied:
They are just bit names concatenated. You dont need a register address as all bit names are connected to a register address. I jsut ensure all these addresses are the same, set the bits and set that value to the register.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        os << "[address] [32 bit "
+              "value][(optional)--validate]\n\t[Mythen3][Gotthard2] "
+              "Reads/writes to a 32 bit register in hex. Advanced "
+              "Function!\n\tGoes to stop server. Hence, can be called while "
```
MMh maybe I would completely seperate bit names from this command and only use it in bitoperations

thattil replied:
but this could also discourage hard coded values into a register. Bit names concatenation would give a clearer idea of what is being set in this register.

thattil replied:
Makes sense. Its not intuitive. Will remove bit names or concatenation of bit names from the reg command

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        }
+
+        // if args0 does not have '|', then get bit definition directly
+        if (sbitNames.find('|') == std::string::npos) {
```
Here again I would seperate bitnames and implement those things in parseBitposition

thattil replied:
parse bit position is only parsing the combination of address name/value + bit position or a bit name.

This concatenation of bit names is only an alternative for writing a value to a register. It is not to parse what the bit position is.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CallerSpecial.cpp:
```diff
@@ -1436,5 +1436,446 @@ std::string Caller::sleep(int action) {
+        }
+
+        auto validate = parseValidate();
+        auto bitPosition = parseBitPosition(validate);
```
again add an INFO if no register definition is used that you set the value at register {adress} to {value}

thattil replied:
Will add at command line level.

AliceMazzoleni99 commented on python/src/detector.cpp:
```diff
@@ -1796,6 +1800,78 @@ void init_det(py::module &m) {
+                       (std::string(Detector::*)(sls::BitPosition) const) &
+                           Detector::getBitDefinition,
+                       py::arg());
+    CppDetectorApi.def("clearBitDefinitions",
```
All the bit operations cant be called as a bit definition name from python. Maybe overload these in detector.py and pass a string e.g. bit name and also declare a dictionary with bit name and value. getting an item then calls getBitDefinition() and then clearBit or setBit depending on the set value.

thattil replied:
You can using d.setBit(d.getBitDefinition('start_trigger_bit')). However, tomorrow afternoon if we decide to implement "define" property in python, this can be implemented directly as well.

AliceMazzoleni99 commented on slsDetectorSoftware/src/CtbConfig.h:
```diff
@@ -1,10 +1,61 @@
+#define CTB_SHMVERSION    0x251031
+
+#define CTB_NAME_LENGTH 32
+
```
maybe a std::map<std::string, RegisterAdress> as registers member would have been easier

AliceMazzoleni99 replied:
Probably just dont see why it doesnt work

thattil replied:
I cant use that in shared memory because it allocates dynamically.

AliceMazzoleni99 replied:
Ah oke got it.

AliceMazzoleni99 commented on slsDetectorSoftware/src/Detector.cpp:
```diff
@@ -2407,37 +2402,24 @@ defs::dacIndex Detector::getDacIndex(const std::string &name) const {
-    auto dettype = getDetectorType().squash();
-    if (dettype != defs::CHIPTESTBOARD && dettype != defs::XILINX_CHIPTESTBOARD)
-        throw RuntimeError("Named dacs only for CTB");
     pimpl->setCtbDacName(i, name);
```
it works for all detectors?

AliceMazzoleni99 replied:
mmh maybe stay consistent and only test for isChipTestBoard on DetectorImplementation side. Can be another PR

thattil replied:
It does. I just moved it to DetectorImpl to hide the implementation details as it anyway calls DetectorImpl. Also because I have more of these checks for the define in DetectorImpl.

AliceMazzoleni99 commented on slsDetectorSoftware/src/Detector.cpp:
```diff
@@ -2566,19 +2502,85 @@ defs::dacIndex Detector::getSlowADCIndex(const std::string &name) const {
+RegisterAddress Detector::getRegisterDefinition(const std::string &name) const {
+    return pimpl->getRegisterDefinition(name);
+}
+
```
maybe these should return a RegisterDefinition or a std::pair<std::string, RegisterAddress> if we use std::unordered_map

thattil replied:
I dont see why it has to be more complicated. It only requires a string  as return type.

AliceMazzoleni99 replied:
I see. For me it is just more explicit if one exposes getRegisterAdress or getRegisterName to Detector and only use those. I found it at first a bit confusing that the funtion returns different things for different input values. But its only a minor suggestion.

thattil replied:
That makes sense actually. The issue is with BitPosition. getBitPosition().bitPosition(). Lets discuss this tomorrow and see what is more intuitive.

AliceMazzoleni99 commented on slsDetectorSoftware/src/DetectorImpl.cpp:
```diff
@@ -1935,86 +1935,240 @@ void DetectorImpl::setBadChannels(const std::vector<int> list, Positions pos) {
+        throw RuntimeError("Bit Definitions only for CTB");
+    return ctb_shm()->getBitPosition(name);
+}
+
```
same here maybe return BitDefinition for both

thattil replied:
I am not sure I agree. BitPosition is simpler. Besides Bit definition is only defined internally for the shared memory structure.

AliceMazzoleni99 commented on slsDetectorSoftware/src/inferAction.cpp:
```diff
@@ -788,6 +788,34 @@ int InferAction::defaultpattern() {
 
+int InferAction::define() {
+
+    if (args.size() == 2) {
```
cant there be more than 2 arguments e.g. define bit register_adress bit_position

thattil replied:
Yes there can be. However, if there are only 2 arguments, it can only be a get.

thattil commented on slsDetectorSoftware/src/inferAction.cpp:
```diff
@@ -788,6 +788,34 @@ int InferAction::defaultpattern() {
+    }
+
+    if (args.size() == 3) {
+        return slsDetectorDefs::PUT_ACTION;
```
But this is incorrect. sls_detector_get define bit 0x200 20. This can be a get. I have to fix this.

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+    }
+} __attribute__((packed));
+
+class BitPosition {
```
Are we happy with the BitPosition name? To me it's not intuitive that it takes an address and position. Maybe BitDefinition/BitAddress or something similar?

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -3,6 +3,7 @@
 #pragma once
 
 #include <bitset>
+#include <stdint.h>
```
Please use `#include <cstdint>` which is the C++ version of the header

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+    constexpr bool operator!=(const RegisterAddress &other) const {
+        return (addr_ != other.addr_);
+    }
+} __attribute__((packed));
```
Do you have a specific reason to use attribute packed? With a single member it doesn't make a different except that it changes the alignment from 4 to 1 bytes.

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+    constexpr bool operator!=(const BitPosition &other) const {
+        return !(*this == other && bitPos_ == other.bitPos_);
+    }
+} __attribute__((packed));
```
Same thing here. With two 4 byte members there should be no gaps and I don't think attribute packed is needed.

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+class BitPosition {
+  private:
+    RegisterAddress addr_{0};
+    int bitPos_{0};
```
Since we use uint32_t in RegisterAddress I would use int32_t for consistency.

thattil replied:
I didnt really get this? We use uint32_t for RegisterAddress and here we use RegisterAddress, so it should be the same right?

AliceMazzoleni99 replied:
I think its because the bitPos can be a value between 0-31 uint8_t makes more sense or why int32_t?

thattil replied:
ah about bitPosition. Got it. Will change to uint32_t as well.

thattil replied:
Done

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+        return (addr_ == other.addr_ && bitPos_ == other.bitPos_);
+    }
+    constexpr bool operator!=(const BitPosition &other) const {
+        return !(*this == other && bitPos_ == other.bitPos_);
```
You can simplify it to: `return !(*this == other);`

erikfrojdh commented on slsSupportLib/tests/test-bit_utils.cpp:
```diff
@@ -42,4 +43,141 @@ TEST_CASE("Get set bits from 523") {
+
+}
+
+TEST_CASE("Implicitly convert to uint for sending over network", "[support][bit_utils]") {
```
I'm a bit lost. Why would we want to convert it to an uint64 and then send it instead of sending the address as a 32 bit value directly?

thattil replied:
Implicit conversion removed

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+    explicit RegisterValue(const std::string &value);
+
+    std::string str() const;
+    operator uint32_t() const noexcept { return value_; }
```
I'm not sure about the conversion operator... I would prefer a `uint32_t RegisterValue::value()` instead. To easy conversion defeats the purpose of having a strong type.

thattil replied:
Ah Makes sense.

thattil replied:
RegisterAddresss and RegisterValue: Removed the implicit conversions.
RegisterAddress: Changed member name from address_ to value_ and method as well to value(). 
RegisterValue: Also added | operator to be able to concatenate with uint32_t. Same in python bindings (but could not find the tests to modify).

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+
+  public:
+    constexpr RegisterAddress() noexcept = default;
+    explicit RegisterAddress(uint32_t address) : addr_(address) {}
```
Can be constexpr

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+    std::string str() const;
+    operator uint32_t() const noexcept { return value_; }
+
+    RegisterValue &operator|=(uint32_t rhs) noexcept {
```
Do we want to allow this with an uint32_t and not only with other RegisterValues?

thattil replied:
Done. Allowed to concatenate with both RegisterValue and uint32_t

erikfrojdh commented on slsSupportLib/include/sls/bit_utils.h:
```diff
@@ -18,4 +19,76 @@ template <typename T> std::vector<int> getSetBits(T val) {
+
+    RegisterAddress address() const noexcept { return addr_; }
+    int bitPosition() const noexcept { return bitPos_; }
+    void setAddress(RegisterAddress address) noexcept { addr_ = address; }
```
Is there a need for the setters? Or we just do it on construction?

thattil replied:
To Modify it? I have only used it  once in a tests though.. we can remove it if its unnecessary..

thattil replied:
Removed.

erikfrojdh reviewed (changes requested):
Added my comments. Mostly very quick fixes.  I have an idea how to make the python part a bit easier to use but will try tomorrow.

erikfrojdh commented:
This could be used in python:
```python
In [17]: regs = types.SimpleNamespace(**d.getRegisterDefinitions())

In [18]: d.reg[regs.something]
Out[18]: [0x0]

```

erikfrojdh commented on python/src/bit.cpp:
```diff
@@ -0,0 +1,55 @@
+        .def(py::init<const std::string &>())
+        .def(py::init<uint32_t>())
+        .def(py::init<const RegisterAddress &>())
+        .def("__repr__", &RegisterAddress::str)
```
For __repr__ do we want it to print RegisterAddress(0x45)? str should probably be only 0x45 for writing to file but the type might be useful when doing interactive stuff

erikfrojdh commented on slsDetectorSoftware/src/CtbConfig.h:
```diff
@@ -1,10 +1,61 @@
+
+class RegisterDefinition {
+  private:
+    char name_[CTB_NAME_LENGTH]{};
```
Why use a macro and not constexpr static?

thattil replied:
Using it for SHMVERSION. For CTB_NAME_LENGTH suing inling constexpr at namespace level as it is used in multiple classes.

thattil committed:
fixing so all tests pass

thattil committed:
removed bit names in reg

thattil committed:
changed BitPosition to BitAddress

erikfrojdh committed:
Using define reg/bit from python (#1344)

* define_bit, define_addr in python. 
* setBit/clearBit takes int or addr

erikfrojdh committed:
Merge branch 'developer' into dev/define_cmd

erikfrojdh committed:
added example using bits

thattil committed:
split define into 2 commands define_reg and define_bit, definelist into 2: definelist_reg and definelist_bit

thattil committed:
Merge remote-tracking branch 'origin/dev/define_cmd' into dev/define_cmd

thattil committed:
allow string for register and bit names in c++ api

thattil committed:
refactor from github comments

thattil committed:
naming refactoring (getRegisterDefnition to retunr name and address specifically

thattil committed:
added marker for 8 cmd tests connected to define, changed macro to static constexpr

thattil committed:
changed bitPosition from int to uint32_t

thattil committed:
got rid of setbitposition and setaddress, instead overloaded constructor to take in strings so that the conversion from string to bit address members, takes place within the class for easy maintainance in case type changes

thattil committed:
Removing implicit conversions:
RegisterAddresss and RegisterValue: Removed the implicit conversions.
RegisterAddress: Changed member name from address_ to value_ and method as well to value().
RegisterValue: Also added | operator to be able to concatenate with uint32_t. Same in python bindings (but could not find the tests to modify

thattil committed:
Allowed concatenation with other RegisterValue, made them all constexpr

thattil committed:
fix a ctbConfig test

thattil committed:
Maponstack works with integration tests, but need unit tests

thattil committed:
tests on mapstack

thattil committed:
forgot to add test

AliceMazzoleni99 commented on python/slsdet/utils.py:
```diff
@@ -141,21 +141,38 @@ def make_ip(arg):
     return _make(arg, _slsdet.MacAddr)
 
+def make_register_address(arg):
+    return _make(arg, _slsdet.RegisterAddress)
```
maybe add this in __init__ file e.g. from .utils import make_register_address otherwise people will for sure not find it and use the struct

thattil replied:
Removed

thattil committed:
fixed ctb tests and FixedString being initialized with gibberish

erikfrojdh commented on python/slsdet/registers.py:
```diff
@@ -1,13 +1,28 @@
+    @element
     def __getitem__(self, key):
+        if isinstance(key, str):
+            key = self._detector.getRegisterDefinition(key)
```
Needs to be updated with `getRegisterDefinitionAddress`

thattil replied:
Done

erikfrojdh commented on python/slsdet/registers.py:
```diff
@@ -1,13 +1,28 @@
 
     def __setitem__(self, key, value):
+        if isinstance(key, str):
+            key = self._detector.getRegisterDefinition(key)
```
Here as well: `getRegisterDefinitionAddress`

erikfrojdh commented on python/slsdet/detector.py:
```diff
@@ -1813,6 +1814,109 @@ def reg(self):
+        self.setRegisterDefinition(name, addr)
+
+
+    def define_bit(self, name, addr, bit):
```
Should maybe also take str for the addr in case someone defined the register first.

thattil replied:
Yes, ive added it

erikfrojdh commented on python/slsdet/detector.py:
```diff
@@ -1813,6 +1814,109 @@ def reg(self):
+
+        #New usage with str or BitAddress
+        if isinstance(bit_or_addr, str):
+            bit_or_addr = self.getBitDefinition(bit_or_addr)
```
Now needs `d.getBitDefinitionAddress` but why is address now added. This is really a bit definition

thattil replied:
Because you can also define with bit name and a bitaddress..

erikfrojdh reviewed (changes requested):
Python API needs updating, and we really need API tests for python =)

AliceMazzoleni99 commented on python/slsdet/detector.py:
```diff
@@ -1813,6 +1814,109 @@ def reg(self):
+
+    def define_bit(self, name, addr, bit):
+        """
+        [Ctb] Define a name for a bit in a register to be used later with setBit/clearBit/getBit
```
lets make these keyword only

thattil committed:
wip, removing parsing from string inside the class RegisterAddress, BitAddress and RegisterValue

thattil committed:
updated python bindings

thattil committed:
fixed bit utils test

thattil committed:
wip. renaming getRegisterDefintiionAddress/Name=>getRegisterAddress/Name and similary for getBitDefinitionAddress/Name

thattil committed:
updated python bindings

thattil committed:
fix tests (format)

thattil committed:
a few python tests added and python bindings corrected

thattil committed:
some more changes for test

thattil committed:
some last changes

thattil committed:
tests for python side

thattil committed:
replaceing str with __str__ for bit.cpp

thattil committed:
repr reimplemented for bit.cpp

thattil committed:
removed make with registerAddress etc

thattil committed:
starting server for tests per session and nor module

thattil committed:
killprocess throws if no process found-> github runs fails, changed to pkill and not throw

thattil committed:
clean shm shouldnt raise, in ci binary not found

thattil committed:
ignoring these tests for CI, which fail on CI because simulators are not generated in CI. This is in another PR, where it should work

thattil committed:
Merge branch 'developer' into dev/define_cmd
